codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -801,6 +810,109 @@ private class ServerHttpProtocolImplGenerator(
+            is CollectionShape -> model.expectShape(memberType.member.target)
+            is MapShape -> model.expectShape(memberType.value.target)
+            else -> UNREACHABLE("unexpected member for prefix headers: $memberType")
+        }
+        writer.ifSet(memberType, memberSymbol, "&output.$memberName") { field ->
+            writer.rustTemplate(
+                """
+                for (k, v) in $field {
+                    use std::str::FromStr;
+                    let header_name = http::header::HeaderName::from_str(
+                            &format!("{}{}", ${binding.locationName.dq()}, &k)
+                        )
+                        .map_err(|err| {
+                            #{SmithyHttpServer}::rejection::Serialize::from(
+                                ${(memberName + " cannot be used as a header name").dq()}
+                            )
+                        })?;
+                    use std::convert::TryFrom;
+                    let header_value = ${headerFmtFun(writer, target, memberShape, "v")};
+                    let header_value = http::header::HeaderValue::try_from(header_value)
+                        .map_err(|err| {
+                            #{SmithyHttpServer}::rejection::Serialize::from(
+                                ${(memberName + " cannot be used as a header value").dq()}
+                            )
+                        })?;
+                    response = response.header(header_name, header_value);
+                }
+                """.trimIndent(),
+                *codegenScope
+            )
+        }
+    }
+
+    /**
+     * Format [member] in the when used as an HTTP header
+     */
+    private fun headerFmtFun(writer: RustWriter, target: Shape, member: MemberShape, targetName: String): String {
+        return when {
+            target.isStringShape -> {
+                if (target.hasTrait<MediaTypeTrait>()) {
+                    val func = writer.format(RuntimeType.Base64Encode(runtimeConfig))
+                    "$func(&$targetName)"
+                } else {
+                    "AsRef::<str>::as_ref($targetName)"
+                }
+            }
+            target.isTimestampShape -> {
+                val index = HttpBindingIndex.of(model)
+                val timestampFormat =
+                    index.determineTimestampFormat(member, HttpBinding.Location.HEADER, protocol.defaultTimestampFormat)
+                val timestampFormatType = RuntimeType.TimestampFormat(runtimeConfig, timestampFormat)
+                "$targetName.fmt(${writer.format(timestampFormatType)})?"
+            }
+            target.isListShape || target.isMemberShape -> {
+                throw IllegalArgumentException("lists should be handled at a higher level")
+            }
+            target.isPrimitive() -> {
+                "encoder.encode()"
+            }
+            else -> throw CodegenException("unexpected shape: $target")
```

Comment:
I am not 100% sure we want the codegen to fail here. I think I agree with bailing out in case of an unexpected shape, but for the list shape we probably want to codegen a `todo!` in rust and let it fail when the crate is built. @david-perez WDYT?

Reply:
BTW I copied this behavior from RequestBindingGenerator so it might be considered in the same way -

https://github.com/awslabs/smithy-rs/blob/36faba8ee6860208b821c384f8a611fbd87f8fae/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/RequestBindingGenerator.kt#L263-L269

Reply:
on the client side parsing lists is quite challenging so we handle it at the `List<T>` level—I'm not sure how you've handled it here (haven't read the PR yet), but I would say if:

- Lists aren't supported: Log a big warning and generate `todo!()`
- Lists are supported and handled elsewhere: `PANIC` or `UNREACHABLE`

Reply:
@crisidev Would be great if you can direct me here

Reply:
@guymguym 

> One question is if we can reuse the code from RequestBindingGenerator.kt where I took the headers generation code to ServerHttpProtocolGenerator.kt.

Yes, we should aim for reuse here since our server implementation just deviates in minor things as far as I can see:

1. The variable name is `output` instead of `input`,
2. The variable name is `response` instead of `builder`,
3. We're mapping directly to `rejection::Serialize` via `http::header::HeaderValue::try_from(&*header_value).map_err` instead of using [`aws_smithy_http::operation::BuildError`](https://docs.rs/aws-smithy-http/latest/aws_smithy_http/operation/enum.BuildError.html); and
4. We're not calling `redactIfNecessary` in case of header serialize error.

We should make `renderHeaders` and `renderPrefixHeaders` from `RequestBindingGenerator.kt` public and reconcile. (4) we need to do as well. (1) and (2) we can parametrize and document the methods so that they take in the variable names (alternatively, for (2), we can rename our `response` to `builder` to simplify a bit). For (3), I suggest we keep using `BuildError` and write a `From` impl [in rejection.rs](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/src/rejection.rs) to put it inside a `Serialize` rejection, so we can keep using `?` to bubble it up.

---

@rcoh lists _are supported_ in `@httpHeader` and `@httpPrefixHeader` like in the client and they are being handled higher via [`listForEach` here](https://github.com/awslabs/smithy-rs/blob/main/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/RequestBindingGenerator.kt#L213); this code block has been copied from there. So throwing `IllegalArgumentException` or `PANIC` or `UNREACHABLE` is correct here; does the codebase have a convention on which to use when?

Reply:
`UNREACHABLE` is intended for cases where if kotlin had exhaustiveness checking, the branch wouldn't be required

`PANIC` is for a bug in the code

Reply:
This is being refactored in #1086. I'll remove my code once rebased. Resolving.